### PR TITLE
Add json column

### LINKED ISSUE #15879: Use daml lf json format in custom views example
Depends on https://github.com/digital-asset/daml/issues/13558
- In the custom-views example, show how to store the contracts `JSON` encoded into a PostgreSQL `jsonb` column and how to query on properties of the contract , possibly add an index for (a) specific field(s) embedded in the `json` column.

### PR DESCRIPTION
**Pull request by chunlokling-da**
fixes https://github.com/digital-asset/daml/issues/15879

### CODE REVIEW

File: src/main/java/com/daml/quickstart/iou/ProjectionRunner.java [hunk 1]
```diff
@@ -68,12 +71,13 @@ public static void main(String[] args) {
             ExecuteUpdate.create(
                     "insert into "
                         + projectionTable.getName()
-                        + "(contract_id, event_id, amount, currency) "
-                        + "values (?, ?, ?, ?)")
+                        + "(contract_id, event_id, amount, currency, json_data) "
+                        + "values (?, ?, ?, ?, ?::json)")
```
Comment (fayi-da) [inline src/main/java/com/daml/quickstart/iou/ProjectionRunner.java]: java sql doesn't have json type so we had to do this cast on the db level

**Review by ray-roestenburg-da — APPROVED** [review]

**Review by S11001001 — APPROVED** [review]

**Comment by ray-roestenburg-da** [comment]
@chunlokling-da would be nice to add an index to inside/for the json data and have a query that performs better in the spring boot app. and add some doc about this

File: src/main/java/com/daml/quickstart/iou/ProjectionRunner.java (hunk 1 above)
Comment (ray-roestenburg-da) [reply src/main/java/com/daml/quickstart/iou/ProjectionRunner.java]: I guess you could look into adding a `Bind.JsValue` to custom-views @fayi-da

File: src/main/resources/db/migration/V20221214201500__Create_Initial_Tables.sql [hunk 2]
```diff
@@ -4,5 +4,6 @@ create table events
     contract_id varchar not null,
     event_id    varchar not null,
     currency    varchar not null,
-    amount      numeric not null
+    amount      numeric not null,
+    json_data   json
```
Comment (ray-roestenburg-da) [inline src/main/resources/db/migration/V20221214201500__Create_Initial_Tables.sql]: @fayi-da @chunlokling-da would be interesting (as an example) to add a GIN index on observers in json_data:

```
CREATE INDEX dataobserversgin ON events USING gin (json_data->'observers');
```

so something like `where json_data->'observers' ? 'does-this-observer-exist'` and get the index used?

**Comment by ray-roestenburg-da** [comment]
@chunlokling-da please check the CI build.

**Comment by chunlokling-da** [comment]
oh the CI failure is due to `com.daml:codegen-json-java_2.13:jar:2.5.0` is not yet publish anywhere yet. I am hoping it will be available when the weekly snapshot release is done this Wednesday

File: src/main/java/com/daml/quickstart/iou/ProjectionRunner.java (hunk 1 above)
Comment (fayi-da) [reply src/main/java/com/daml/quickstart/iou/ProjectionRunner.java]: We did this actually but the only solution possible will have that JsValue map to varchar/text in the sql types.

Comment (ray-roestenburg-da) [reply src/main/java/com/daml/quickstart/iou/ProjectionRunner.java]: @fayi-da That's fine, it just means that the user does not have to do 

```
.bind(5, jsonCodec.toJsValue(iou.data.toValue()).compactPrint(), Bind.String());
```

But instead can do

```
.bind(5, jsonCodec.toJsValue(iou.data.toValue()), Bind.JsValue());
```
Or maybe even we could do: (with 'implicit' json encoding, no pun intended)

```
.bind(5, iou.data, Bind.Value());
```

Which is very convenient for the user. The `Bind.Value` bind does the `jsonCodec.toJsValue(_.toValue())`

Comment (ray-roestenburg-da) [reply src/main/java/com/daml/quickstart/iou/ProjectionRunner.java]: So the naming might be a little confusion there, which can be improved, but that would be the idea.

File: src/main/resources/db/migration/V20221214201500__Create_Initial_Tables.sql (hunk 2 above)
Comment (chunlokling-da) [reply src/main/resources/db/migration/V20221214201500__Create_Initial_Tables.sql]: pushed commit to add the gin index and also API to get events filtered by observers field in the json column

Comment (chunlokling-da) [reply src/main/resources/db/migration/V20221214201500__Create_Initial_Tables.sql]: Also tested to make sure the new index is being used when doing `where json_data->'observers' ? 'does-this-observer-exist'`

```
SET enable_seqscan TO off;
EXPLAIN select * from events where json_data->'observers' ? 'EUR_Bank::12201f8d3d6fe6285cc964ac20fdceea91f57c0742abf3ed452e88c0897cb9e3cbfd';
```

```
Bitmap Heap Scan on events  (cost=12.03..20.50 rows=4 width=160)
  Recheck Cond: ((json_data -> 'observers'::text) ? 'EUR_Bank::12201f8d3d6fe6285cc964ac20fdceea91f57c0742abf3ed452e88c0897cb9e3cbfd'::text)
  ->  Bitmap Index Scan on json_data_observers_gin  (cost=0.00..12.03 rows=4 width=0)
        Index Cond: ((json_data -> 'observers'::text) ? 'EUR_Bank::12201f8d3d6fe6285cc964ac20fdceea91f57c0742abf3ed452e88c0897cb9e3cbfd'::text)

```

File: src/main/resources/db/migration/V20221214201500__Create_Initial_Tables.sql [hunk 3]
```diff
@@ -4,5 +4,8 @@ create table events
     contract_id varchar not null,
     event_id    varchar not null,
     currency    varchar not null,
-    amount      numeric not null
+    amount      numeric not null,
+    json_data   jsonb
```
Comment (S11001001) [inline src/main/resources/db/migration/V20221214201500__Create_Initial_Tables.sql]:
```suggestion
    json_data   jsonb not null
```

File: src/main/java/com/daml/quickstart/iou/ProjectionRunner.java (hunk 1 above)
Comment (S11001001) [reply src/main/java/com/daml/quickstart/iou/ProjectionRunner.java]: It's exactly the kind of thing `contramap` is good for.

**Review by S11001001 — APPROVED** [review]

**Comment by ray-roestenburg-da** [comment]
It's likely needed to depend on a snapshot @chunlokling-da @fayi-da ?

**Comment by chunlokling-da** [comment]
Yes. The required PR in daml is merged: https://github.com/digital-asset/daml/pull/15909
So we can put the released snapshot version this Wed or Thur

**Pull request merged**